src/style/style_layer/line_style_layer.js

```diff
@@ -50,18 +50,32 @@ class LineStyleLayer extends StyleLayer {
     _transitionablePaint: Transitionable<PaintProps>;
     _transitioningPaint: Transitioning<PaintProps>;
     paint: PossiblyEvaluated<PaintProps>;
+    transitionParameters: EvaluationParameters;
 
     constructor(layer: LayerSpecification) {
         super(layer, properties);
     }
 
     recalculate(parameters: EvaluationParameters) {
         super.recalculate(parameters);
+        this.transitionParameters = parameters;
 
         (this.paint._values: any)['line-floorwidth'] =
             lineFloorwidthProperty.possiblyEvaluate(this._transitioningPaint._values['line-width'].value, parameters);
     }
 
+    getCrossfadeParameters(): {fromScale: number, toScale: number, t: number} {
+        const p = this.transitionParameters;
+        const z = p.zoom;
+        const fraction = z - Math.floor(z);
+        const t = p.crossFadingFactor();
+
+        return z > p.zoomHistory.lastIntegerZoom ?
+            { fromScale: 2, toScale: 1, t: fraction + (1 - fraction) * t } :
+            { fromScale: 0.5, toScale: 1, t: 1 - (1 - t) * fraction };
+    }
```

Comment:
When I implement `fill-pattern` and `fill-extrusion-pattern` dds in a follow-up PR, I will probably move this function to the parent `StyleLayer` class.